### DISCUSSION TITLE
Add Notion-like inline databases

### DISCUSSION THREAD

**Discussion by SaadBazaz** [outline/outline#6507] [Ideas]
Hey,
The only reason which prevents me from totally yeeting Notion out the window is the "inline database" feature, which enables one to maintain data in many creatives ways.

[AFFiNE](https://affine.pro/) does this, however their self-hosted version is hardly in an alpha stage, if even anything at all.

If we can get this (even a third-party plugin works!), we're set!

**Comment by SaadBazaz** [comment]
Followup to this:

If we could attach NocoDB as a plugin, and in every row of the table, add a hidden column which links to an Outline page, we could have a bidirectional sync between Outline and NocoDB.

I think Outline + NocoDB can be an amazing combo to achieve fun, creative Databases

**Reply by HugoooPerrin** [reply]
This feature would be incredible. It would indeed make Outline relevant for many more usecases!

I saw in the docs that there is an embedding for Airtable databases, would it be so different for NocoDB ?

**Reply by SaadBazaz** [reply]
There's Airtable, but that's more like an embed/webview/iframe. Sure, it's nice to have, but I'm talking at least one step more deeper than that. e.g. actually storing the page's ID in the table, so it's searchable, etcetera.

So a bit more customised, rather than just an embed.

**Reply by HugoooPerrin** [reply]
Ok I see, even more on board then!

**Comment by SaadBazaz** [comment]
Further followup...

## Nocodb + Outline

I'm thinking of a complete Notion alternative using self-hosted projects. I think the ingredients are all there, if @getoutline collaborates with @nocodb, we can achieve complex Inline Databases within Outline. No need to add unnecessary feature creep, just reuse from another project!

<img width="1512" alt="Screenshot 2024-02-13 at 09 34 39" src="https://github.com/outline/outline/assets/51885228/71ef317d-1b4c-4348-a3ff-4c27b1848adb">

I'll see if I can design a UI for this.

**Reply by sabgaby** [reply]
This looks great!, I feel like Plane is missing a bit of text editing that could be implemented with Outline. In terms of database, why NocoDB and not [Teable](https://teable.io/) or [Appflowy](https://www.appflowy.io/). Just curiosity on your preference.

**Reply by SaadBazaz** [reply]
@sabgaby AirFlowy is very app-like, not a very pleasant web experience (I'm probably biased, given that Appflowy is Flutter). Outline and Plane have almost the same stack to begin with (Both are NextJS apps, both use Django as backend, and similar databases), so it seems easier to mix them both together.

Treable looks nice, but Nocodb really is an established player (so far)

**Reply by niscolas** [reply]
During some research I've also found the well established [Baserow](https://gitlab.com/baserow/baserow) and this newer one [APITable](https://github.com/apitable/apitable)

**Reply by SaadBazaz** [reply]
Thanks @niscolas . I think Outline already supports [Grist](https://www.getgrist.com/) too, but which low-code DB module is less of a point. The point is that ANY good enough low-code DB should have some tighter integration with Outline. So we can be more creative with documents.

**Reply by Novapixel1010** [reply]
Little late to the party, I think [Baserow](https://gitlab.com/baserow/baserow) and [NocoDB](https://github.com/nocodb/nocodb) are both good options. They are both open source and seem good. I like **Baserow** more because it is MIT licensed but that is just a personal thing.

**Comment by nickcv-ln** [comment]
this would actually be amazing
it is really the only major feature missing from notion

**Comment by pmdevita** [comment]
Definitely would be nice to see, is the Outline team considering this for their roadmap?

**Reply by almereyda** [reply]
I'd love to see this play out in a decoupled way through an increasingly intimate integration with Grist. It's basically the same idea as above, but swapping out NocoDB with a low code database that is already integrated.

With which I held a suprising conversation lately. https://fosstodon.org/@grist/112651430651558497

---

There's even place for a triple with a #6407-like integration for Plane.

Work is ongoing in their community to embed Outline (collections?) into Plane projects, replacing their Pages feature. Please find this PR:

- https://github.com/makeplane/plane/pull/4366

> *Edit:* I wrote this as a response to the notification, before I scrolled back up to reconstruct the context. It means the idea has come up twice independently.

In a similar manner of such a reverse integration, it is possible to consider integrating outline documents in Grist widgets. The setup and editing UX of the Markdown widget is close to non-existing. AnyType, AppFlowy and LogSeq show that it is possible to get close to the Notion ideal with an open implementation. Also the Mattermost Focalboard, before it was "donated" to "the community" (abandoned), showed an implementation that was going into the right directions.

Maybe time projects lend each other hands by covering each others backs in certain regard? This perspective assumes that there is another purpose to social organisation at play when businesses don't produce Software to generate income, but also to contribute to a Common.

As a not-for-profit, we are a generous user of the in-two-years FLOSS edition of Outline, but today, which is a compromise we can take.

---

The Nextcloud–OpenProject integration was an interesting example in setting up a bidirectional link between two platforms in a runtime environment.

We might see more of these closer integrations between decoupled platforms, now that the web is starting to learn to federate properly (GitLab, Forgeo, WordPress, Ghost, Discourse, OIDC self-registration, OIDC federation, Federated Wiki and [ibis.wiki](https://ibis.wiki/)). Exciting times ahead.

**Reply by almereyda** [reply]
Teable now has support for versioned records and it is growing well in the past. In contrast to Grist, it is fast by default, as it uses PostgreSQL and not SQLite-on-S3.

The need for inlining block editors with structured data is also prevalent from the side of their community, where Outline has been explicitly mentioned as Notion alternative.

Find out more at:

- https://github.com/teableio/teable/issues/829

**Reply by D1reW0lf** [reply]
> Teable now has support for versioned records and it is growing well in the past. In contrast to Grist, it is fast by default, as it uses PostgreSQL and not SQLite-on-S3.
> 
> The need for inlining block editors with structured data is also prevalent from the side of their community, where Outline has been explicitly mentioned as Notion alternative.
> 
> Find out more at:
> 
> * [[Plugin] Block editor view teableio/teable#829](https://github.com/teableio/teable/issues/829)

Indeed, a proper teable integration would be great. 
Teable is coming along really nice with bidirectional links between databases, among other things.
Outline + Teable on the same "document" would be the best of both worlds.

**Reply by almereyda** [reply]
And vice versa.

**Comment by notfounnd** [comment]
Some news about this idea on Outline roadmap?

**Reply by SaadBazaz** [reply]
I'm assembling a team to get this done. Will ping here if any progress.

**Reply by Atomique** [reply]
Any news here?

**Comment by ghost** [comment]
Hey everyone! I’ve started a new discussion on adding **structured data (databases)** to Outline and referenced this thread among other related ideas.

The new post is quite visual and includes a current ecosystem overview, potential features, and mockups. I’d love your feedback:
⭐ #8476

**Comment by Plinsboorg** [comment]
Hi all. I've implemented a draft version of the document databases for Outline.
PR: https://github.com/outline/outline/pull/13210
Try it: https://outline-db-test.plinsburg.tech/
the magic email sign in will let you in 
<img width="1873" height="690" alt="image" src="https://github.com/user-attachments/assets/f186ab4d-9e5e-4301-8e5a-8475a5a4f92e" />

See this discussion for details
https://github.com/outline/outline/discussions/8476#discussioncomment-17851977

I'd love to hear from maintainers if this fully vibe coded solution has a chance to be accepted.

**Reply by TGlide** [reply]
No one wants to review a 15k LOC PR, and do the work for you of checking if its good. Its a waste of maintainers time (and your tokens).

**Reply by tarmacx** [reply]
I totally get where you're coming from @TGlide. 15k lines of code is a nightmare to review and no one expects you guys to sift through all that.

That said, @Plinsboorg is clearly excited about this and willing to put in the work. It would be awesome to see this feature in Outline (I myself stumbled into this issue while searching for that feature). If the team is open to the idea itself, could you point us in the right direction on how to break this down properly? What would be the best way to stage and structure smaller PRs so it’s actually manageable for you to look at?

@Plinsboorg, definitely a good lesson for next time 😄 always open an issue first to pitch the data model and tech choices before writing any code or burning tokens. Let's see if we can salvage this by splitting it up into bite-sized pieces. An LLM can produce a LOT of work, but it takes a LOT of work to review it and integrate it properly.

**Reply by Plinsboorg** [reply]
This feature is quite deep. I started using it on my server and I see how it can become an endless feature request generator. Some subtle things like resize and reorder the columns, then drag and drop, sort, filter, it all takes extra code, and I don't quite see the end of it yet. 
Still I need a solution for myself so either I keep improving this, or I need to look around for another project that already does it. 
I'd be open to splitting the work in the smaller pieces but I also get how the outline developers may want to keep this feature to an enterprise plan anyway. 
For anyone interested, you're welcome to give a try to my fork. The implementation is on the feature branch.

**Reply by tarmacx** [reply]
Have you tough about maybe creating an integration/extension instead of a new type of document ? That would lower the point of contact between your feature and the main software?
There are some examples of in the code that you could check like diagrams.net and others.
There's even one for Airtable which could achieve something similar than what you want to do.
I know Airtable is paid but what about creating an extension for NocoDB or Baserow instead of building your own services. That way the plugin/integration itself is a very simple code, easy to review and merge then all the magic happens in NocoDB or Baserow which could be launched as a side service locally or use their cloud offering which is in the same mindset as Outline.

You could also explore Docmost implementation of bases in their ee license to have an idea on how they approach this requirement if you persist in that direction.

And as for your requirement, I'm currently looking for the exact same thing and I still picked Outline instead of Docmost beacause of the pricing structure and featureset, I havent found something "like" Outline but with db outside of this and even Docmost loose features compared to Outline and adds some other mismatch for my requirements.

I hope you'll find a way to make your feature acceptable by the community as this would fill the only missing feature Outline has for me.

Regards

**Reply by TGlide** [reply]
I'm not a maintainer :sweat_smile: 

I don't see a problem with a fork though. My comment was more to say: big features like these are rarely tackled by first-time contributors though, specially without initial discussion with the maintainers. Its too hard to coordinate that way.

But its good that the solution works for you.

**Reply by Plinsboorg** [reply]
I tried Affine, and I'm not a fan of not having documents all sitting in a structured way on the left. 
I don't know any other self-hostible alternative, so I went back to Outline
@TGlide I've created a fork. Well, Claude did it. 
https://github.com/Plinsboorg/outline
Here is the brakedown of new lines of code added per feature
<img width="814" height="425" alt="image" src="https://github.com/user-attachments/assets/6cd1b5f8-f34f-4110-99a5-9bbe54dc4065" />

You're welcome to give this a try, and keep improving the feature with me. 
Make sure to backup your data if you want to use it on "production" as I do.
At the point where it becomes stable and usable to me, I'll try to chunk it up, and submit upstream piece by piece if maintainer will be interested. 
At this point I'm just solving this problem for myself
